libp2p/protocols/rendezvous.nim

```diff
@@ -0,0 +1,648 @@
+        ))
+  await conn.writeLp(msg.buffer)
+
+proc sendDiscoverResponseError(conn: Connection,
+                               status: ResponseStatus,
+                               text: string = "") {.async.} =
+  let msg = encode(Message(
+        msgType: MessageType.DiscoverResponse,
+        discoverResponse: some(DiscoverResponse(status: status, text: some(text)))))
+  await conn.writeLp(msg.buffer)
+
+proc countRegister(rdv: RendezVous, peerId: PeerId): int =
+   for data in rdv.registered:
+     if data.peerId == peerId:
+       result.inc()
+
+proc save(rdv: RendezVous,
+          ns: string,
+          peerId: PeerId,
+          r: Register,
+          update: bool = true) =
+  let nsSalted = ns & rdv.salt
+  discard rdv.indexes.hasKeyOrPut(nsSalted, newSeq[int]())
+  try:
+    for index in rdv.indexes[nsSalted]:
+      if rdv.registered[index].peerId == peerId:
+        if update == false: return
+        rdv.registered[index].expiration = rdv.defaultDT
+    rdv.registered.add(
+      RegisteredData(
+        peerId: peerId,
+        expiration: Moment.now() + r.ttl.get(MinimumTTL).int64.seconds,
+        data: r
+      )
+    )
+    rdv.indexes[nsSalted].add(rdv.registered.high)
+#    rdv.registerEvent.fire()
+  except KeyError:
+    doAssert false, "Should have key"
+
+proc register(rdv: RendezVous, conn: Connection, r: Register): Future[void] =
+  trace "Received Register", peerId = conn.peerId, ns = r.ns
+  if r.ns.len notin 1..255:
+    return conn.sendRegisterResponseError(InvalidNamespace)
+  let ttl = r.ttl.get(MinimumTTL)
+  if ttl notin MinimumTTL..MaximumTTL:
+    return conn.sendRegisterResponseError(InvalidTTL)
+  let pr = checkPeerRecord(r.signedPeerRecord, conn.peerId)
+  if pr.isErr():
+    return conn.sendRegisterResponseError(InvalidSignedPeerRecord, pr.error())
+  if rdv.countRegister(conn.peerId) >= RegistrationLimitPerPeer:
+    return conn.sendRegisterResponseError(NotAuthorized, "Registration limit reached")
+  rdv.save(r.ns, conn.peerId, r)
+  conn.sendRegisterResponse(ttl)
+
+proc unregister(rdv: RendezVous, conn: Connection, u: Unregister) =
+  trace "Received Unregister", peerId = conn.peerId, ns = u.ns
+  let nsSalted = u.ns & rdv.salt
+  try:
+    for index in rdv.indexes[nsSalted]:
+      if rdv.registered[index].peerId == conn.peerId:
+        rdv.registered[index].expiration = rdv.defaultDT
+  except KeyError:
+    return
+
+proc discover(rdv: RendezVous, conn: Connection, d: Discover) {.async.} =
+  trace "Received Discover", peerId = conn.peerId, ns = d.ns
+  if d.ns.len notin 0..255:
+    await conn.sendDiscoverResponseError(InvalidNamespace)
+    return
+  var limit = min(DiscoverLimit, d.limit.get(DiscoverLimit))
+  let
+    nsSalted = d.ns & rdv.salt
+    cookie =
+      if d.cookie.isSome():
+        try:
+          Cookie.decode(d.cookie.get()).get()
+        except CatchableError:
+          await conn.sendDiscoverResponseError(InvalidCookie)
+          return
+      else: Cookie()
```

Comment:
I think we said at some point that we should start with a random cookie to avoid that everyone gets the same peers?

Reply:
Or maybe we can just shuffle the response, but meh